### PR TITLE
Feature/checkout

### PR DESCRIPTION
**Pull request by TylerRockwell**
### Why?
https://github.com/smashingboxes/apprenticeship/tree/master/Curriculum/Backend/bookstore
Because users should always have a cart when they enter the store
Because cart items need to move to an order when they are purchased

### What Changed?
A cart is now created for a user when their account is created
Creating an order now creates order_items and removes line_items from the cart

### CODE REVIEW

File: app/controllers/orders_controller.rb [hunk 1]
```diff
@@ -0,0 +1,11 @@
+class OrdersController < ApplicationController
+  def create
+    @order = Order.new
+    @order.add_items_from(cart)
```
Comment (jahammo2) [inline app/controllers/orders_controller.rb]: Very easy to read, gj

Comment (TylerRockwell) [inline app/controllers/orders_controller.rb]: Thanks. I was happy with how that method came out.

File: app/models/order.rb [hunk 2]
```diff
@@ -10,6 +10,16 @@ def calculate_total
     order_items.inject(0) { |sum, item| sum + item.total_price }
   end
 
+  def add_items_from(cart)
```
Comment (albaer) [inline app/models/order.rb]: Looks great!

File: features/step_definitions/order_steps.rb [hunk 3]
```diff
@@ -0,0 +1,19 @@
+When(/^I enter my payment details$/) do
+  sleep(2) # wait for the js to create the popup in response to pressing the button
```
Comment (albaer) [inline features/step_definitions/order_steps.rb]: Hahaha. The 0.7 wasn't cutting it?

Comment (TylerRockwell) [inline features/step_definitions/order_steps.rb]: No, my system is too slow :cry:.  It was only passing about half the time.

File: app/models/user.rb [hunk 4]
```diff
@@ -1,9 +1,16 @@
...
+  has_many :addresses
   has_one  :cart
+
+  after_create :build_cart
+
+  private
+
+  def build_cart
+    cart || Cart.create(user: self)
+  end
```
Comment (jahammo2) [inline app/models/user.rb]: Interesting way of utilizing a cart. I like it.

File: features/scenarios/order.feature [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+    Given I am logged into the site
+      And there are some books in the database
+    When I visit the public book index
+      And I click on a book
+      And I click "Add to Cart"
+      And I visit my cart
+      And I click "Purchase"
+      And I click "Pay with Card"
+      And I enter my payment details
+    Then I see "Your order has been placed"
```
Comment (jahammo2) [inline features/scenarios/order.feature]: I have not indented the `And`s before but I think I will now. This definitely makes it very readable.

File: features/step_definitions/order_steps.rb (hunk 3 above)
Comment (jahammo2) [inline features/step_definitions/order_steps.rb]: Good use of a comment

**Comment by jahammo2** [comment]
This is really good. :shipit:

File: features/scenarios/order.feature (hunk 5 above)
Comment (TylerRockwell) [inline features/scenarios/order.feature]: Cool. I thought so too.

**Comment by TylerRockwell** [comment]
Thanks, @jahammo2 and @albaer !

**Pull request merged**